src/olympia/devhub/utils.py

```diff
@@ -183,6 +184,51 @@ def _merged_messages():
     }
 
 
+def find_previous_version(addon, file, version):
+    """
+    Find the most recent previous version of this add-on, prior to
+    `version`, that can be used to compare validation results or
+    issue upgrade warnings.
+    """
+    if not addon:
+        return
+
+    version = Version(version)
+    # Find any previous version of this add-on with the correct status
+    # to match the given file.
+    files = File.objects.filter(version__addon=addon)
+
+    if addon.is_listed and (file and file.status != amo.STATUS_BETA):
+        # TODO: We'll also need to implement this for FileUploads
+        # when we can accurately determine whether a new upload
+        # is a beta version.
+        files = files.filter(status=amo.STATUS_PUBLIC)
+    else:
+        files = files.filter(Q(status=amo.STATUS_PUBLIC) |
+                             Q(status=amo.STATUS_BETA, is_signed=True))
+
+    if file:
+
+        # Add some extra filters if we're validating a File instance,
+        # to try to get the closest possible match.
+        files = (files.exclude(pk=file.pk)
+                 # Files which are not for the same platform, but have
+                 # other files in the same version which are.
+                 .exclude(~Q(platform=file.platform) &
+                          Q(version__files__platform=file.platform))
+                 # Files which are not for either the same platform or for
+                 # all platforms, but have other versions in the same
+                 # version which are.
+                 .exclude(~Q(platform__in=(file.platform,
```

Comment:
Yeah, I agree. But to be honest, I'm scared touching this since it appears to be working the way it is (tm)